labs/b9.md

```diff
@@ -0,0 +1,178 @@
+
+For our last trick, we’re going to use Docker to run multiple Apache web servers inside containers. 
+
+For simplicity, you will not have to write this `Dockerfile`. Go ahead and pull the `httpd` image from Docker Hub. Now, it’s your job to figure out how to run three instances of the Apache containers on your machine.           
+
+Docker creates a separate network for containers, so you will need to forward your host port to your container’s port (this is called [port forwarding](https://en.wikipedia.org/wiki/Port_forwarding), or port mapping). The container is listening on port 80 by default. It is your job to run each instance on ports 5000, 5001, and 5002.  I recommend running the containers in detached mode with the `-d` flag.
+
+Hints (if you want them): 
+- Don’t expect to see any output until you visit the web application from your host machine.
+- The `-p` flag takes in a colon separated pair of HOST\_PORT:CONTAINER\_PORT (it can actually accept a ton of more options, but don’t worry about that for now). 
+- You should be able to view the web application from your host machine by visiting <url_of_host_machine>:5000, assuming you don’t have anything else running on that port.      
+ 
+ ##### Questions
+ 1. While your three containerized Apache web servers are running in detached mode, paste the output of `docker ps`.
+ 2. Observe that in the output of `docker ps` each container has an associated container ID. Explain why containers have IDs rather than being named after the application, for example httpd. 
+ 3. Now go ahead and stop your containers (see slide 30). Paste the command you used to stop one of the containers. 
+
+Congratulations! You’ve successfully Dockerized and ran a web application with vastly different requirements and configuration than your host machine, without affecting your setup on your machine :) There’s a lot more about Docker and containers to learn about, but I hope this was enough to wrap your head around the basic concepts and get some experience working with it.        
+
+For further reading, I recommend just reading the [official documentation](https://docs.docker.com/) so you can see what is possible with the Docker container format.    
+
+#### Puppet
+
+To get some hands on experience with Puppet, we are going to write a very basic manifest that pulls quotes from the internet and appends them to a file every two minutes. Although you were taught that in production environments we have the puppet master running on its own server, for simplicity you will apply the manifest locally, making you both the puppet master and the puppet agent. If you are confused about the Puppet vocabulary, check out slide 21.
+
+First things first. We plan to pull quotes from the web, so we need to ensure the `wget` package is installed on our system. Next we need to create a user, 'quotes', that runs this command for us. Take note of the dependency between the 'quotes' user and the 'quotegather' group existing. 
```

Comment:
I'd explain a bit more why this dependency is there, because from this I don't think it's clear. Also I think explaining that running it as a separate user is good for security because the default user is `root` would be a good idea.